Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM.cpp

```diff
@@ -416,29 +416,58 @@ void FiniteDifferenceSolver::MacroscopicEvolveHMCartesian(
         Box const &tby = mfi.tilebox(Hfield[1]->ixType().toIntVect());
         Box const &tbz = mfi.tilebox(Hfield[2]->ixType().toIntVect());
 
-        amrex::Real mu0_inv = 1. / PhysConst::mu0;
+        // read in Ms to decide if the grid is magnetic or not
+        auto& mag_Ms_mf = macroscopic_properties->getmag_Ms_mf();
+        Array4<Real> const& mag_Ms_arr = mag_Ms_mf.array(mfi);
+
+        // mu_mf will be imported but will only be called at grids where Ms == 0
+        auto& mu_mf = macroscopic_properties->getmu_mf();
+        Array4<Real> const& mu_arr = mu_mf.array(mfi);
+
+        amrex::Real const mu0_inv = 1. / PhysConst::mu0;
 
         // Loop over the cells and update the fields
         amrex::ParallelFor(tbx, tby, tbz,
             [=] AMREX_GPU_DEVICE(int i, int j, int k) {
-                Hx(i, j, k) += mu0_inv * dt * T_Algo::UpwardDz(Ey, coefs_z, n_coefs_z, i, j, k)
-                             - mu0_inv * dt * T_Algo::UpwardDy(Ez, coefs_y, n_coefs_y, i, j, k);
-                if (coupling == 1) {
-                    Hx(i, j, k) += - M_xface(i, j, k, 0) + M_xface_old(i, j, k, 0);
+                Real mag_Ms_arrx = MacroscopicProperties::macro_avg_to_face(i, j, k, amrex::IntVect(1,0,0), mag_Ms_arr);
+                if (mag_Ms_arrx == 0._rt){ // nonmagnetic region
+                    Real mu_arrx = MacroscopicProperties::macro_avg_to_face(i, j, k, amrex::IntVect(1,0,0), mu_arr);
+                    Hx(i, j, k) += 1. / mu_arrx * dt * T_Algo::UpwardDz(Ey, coefs_z, n_coefs_z, i, j, k)
+                                 - 1. / mu_arrx * dt * T_Algo::UpwardDy(Ez, coefs_y, n_coefs_y, i, j, k);
+                } else if (mag_Ms_arrx > 0){ // magnetic region
```

Comment:
looks like there is nothing for Ms <0 . So in my previous comment, may be instead of writing the condition as `Ms!=0`, we could so `Ms > 0` ?